WalletWasabi/Services/SystemAwakeChecker.cs

```diff
@@ -1,26 +1,100 @@
+using WalletWasabi.Helpers.PowerSaving;
+using WalletWasabi.Logging;
 using WalletWasabi.Wallets;
+using static WalletWasabi.Helpers.PowerSaving.LinuxInhibitorTask;
 
 namespace WalletWasabi.Services
 {
 	public class SystemAwakeChecker : PeriodicRunner
 	{
-		public SystemAwakeChecker(WalletManager walletManager) : base(TimeSpan.FromMinutes(1))
+		private const string Reason = "CoinJoin is in progress.";
+		private static readonly TimeSpan Timeout = TimeSpan.FromMinutes(5);
+
+		private volatile IPowerSavingInhibitorTask? _powerSavingTask;
+
+		private SystemAwakeChecker(WalletManager walletManager, Func<Task<IPowerSavingInhibitorTask>>? taskFactory) : base(TimeSpan.FromMinutes(1))
 		{
 			WalletManager = walletManager;
+			TaskFactory = taskFactory;
 		}
 
 		private WalletManager WalletManager { get; }
+		public Func<Task<IPowerSavingInhibitorTask>>? TaskFactory { get; }
+
+		/// <summary>Checks whether we support awake state prolonging for the current platform.</summary>
+		public static async Task<SystemAwakeChecker?> CreateAsync(WalletManager walletManager)
+		{
+			Func<Task<IPowerSavingInhibitorTask>>? taskFactory = null;
+
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+			{
+				// Either we have systemd or not.
+				bool isSystemd = await LinuxInhibitorTask.IsSystemdInhibitSupportedAsync().ConfigureAwait(false);
+				if (!isSystemd)
+				{
+					return null;
+				}
+
+				GraphicalEnvironment gui = GraphicalEnvironment.Other;
+
+				// Specialization for GNOME.
+				if (await LinuxInhibitorTask.IsMateSessionInhibitSupportedAsync().ConfigureAwait(false))
+				{
+					gui = GraphicalEnvironment.Mate;
+				}
+				else if (await LinuxInhibitorTask.IsGnomeSessionInhibitSupportedAsync().ConfigureAwait(false))
+				{
+					gui = GraphicalEnvironment.Gnome;
+				}
+
+				taskFactory = () => Task.FromResult<IPowerSavingInhibitorTask>(LinuxInhibitorTask.Create(InhibitWhat.Idle, Timeout, Reason, gui));
+			}
+
+			return new SystemAwakeChecker(walletManager, taskFactory);
+		}
 
 		protected override async Task ActionAsync(CancellationToken cancel)
 		{
+			IPowerSavingInhibitorTask? task = _powerSavingTask;
+
 			if (WalletManager.AnyCoinJoinInProgress())
 			{
-				await EnvironmentHelpers.ProlongSystemAwakeAsync().ConfigureAwait(false);
+				if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+				{
+					if (task is not null)
+					{
+						if (!task.Prolong(Timeout.Add(TimeSpan.FromMinutes(1))))
+						{
+							Logger.LogTrace("Failed to prolong the power saving task.");
+							task = null;
+						}
+					}
+
+					if (task is null)
+					{
+						Logger.LogTrace("Create new power saving prevention task.");
+						_powerSavingTask = await TaskFactory!().ConfigureAwait(false);
+					}
+				}
+				else
+				{
+					await EnvironmentHelpers.ProlongSystemAwakeAsync().ConfigureAwait(false);
+				}
+			}
+			else
+			{
+				if (task is not null)
+				{
+					Logger.LogWarning("Stop the power saving task.");
```

Comment:
The message is misleading. Wasabi saving power?

Reply:
So AFAIK people call APIs for prolonging awake state "power saving APIs" as it modifies power saving options / schemas. I don't really know about a better name. Here, it does not make too much sense as the sentence should say something like "Computer idle state is allowed again." or something. Not sure about proper wording.